aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move

```diff
@@ -1,97 +1,160 @@
-    // public fun test_string_concat1(fx: &signer) {
-    //     use std::features;
-    //     let feature = features::get_aggregator_snapshots_feature();
-    //     features::change_feature_flags(fx, vector[feature], vector[]);
-
-    //     let snapshot = create_snapshot(42);
-    //     let snapshot2 = string_concat(std::string::utf8(b"before"), &snapshot, std::string::utf8(b"after"));
-    //     assert!(read_snapshot(&snapshot2) == std::string::utf8(b"before42after"), 0);
-    // }
-
-    // #[test(fx = @std)]
-    // public fun test_string_concat2(fx: &signer) {
-    //     use std::features;
-    //     let feature = features::get_aggregator_snapshots_feature();
-    //     features::change_feature_flags(fx, vector[feature], vector[]);
-
-    //     let snapshot = create_snapshot<String>(std::string::utf8(b"42"));
-    //     let snapshot2 = string_concat(std::string::utf8(b"before"), &snapshot, std::string::utf8(b"after"));
-    //     assert!(read_snapshot(&snapshot2) == std::string::utf8(b"before42after"), 0);
-    // }
-
-    // #[test]
-    // #[expected_failure(abort_code = 0x030006, location = Self)]
-    // public fun test_snapshot_feature_not_enabled() {
-    //     create_snapshot(42);
-    // }
-
-    // #[test(fx = @std)]
-    // #[expected_failure(abort_code = 0x030005, location = Self)]
-    // public fun test_snpashot_invalid_type1(fx: &signer) {
-    //     use std::features;
-    //     use std::option;
-    //     let feature = features::get_aggregator_snapshots_feature();
-    //     features::change_feature_flags(fx, vector[feature], vector[]);
-
-    //     create_snapshot(option::some(42));
-    // }
-
-    // #[test(fx = @std)]
-    // #[expected_failure(abort_code = 0x030005, location = Self)]
-    // public fun test_snpashot_invalid_type2(fx: &signer) {
-    //     use std::features;
-    //     let feature = features::get_aggregator_snapshots_feature();
-    //     features::change_feature_flags(fx, vector[feature], vector[]);
-
-    //     create_snapshot(vector[42]);
-    // }
+    /// Concatenates `before`, `snapshot` and `after` into a single string.
+    /// snapshot passed needs to have integer type - currently supported types are u64 and u128.
+    /// raises EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE if called with another type.
+    public native fun string_concat<IntElement>(before: String, snapshot: &AggregatorSnapshot<IntElement>, after: String): AggregatorSnapshot<String>;
+
+    #[test(fx = @std)]
+    public fun test_correct_read(fx: &signer) {
+        use std::features;
+        let feature = features::get_aggregator_snapshots_feature();
+        features::change_feature_flags(fx, vector[feature], vector[]);
+
+        let snapshot = create_snapshot(42);
+        let snapshot2 = copy_snapshot(&snapshot);
+        assert!(read_snapshot(&snapshot) == 42, 0);
+        assert!(read_snapshot(&snapshot2) == 42, 0);
+    }
+
+    #[test(fx = @std)]
+    public fun test_correct_read_string(fx: &signer) {
```

Comment:
Do these tests succeed now? They didn't work last time when I tried to merge aggregator snapshots into main.